### ISSUE TITLE
License: Apache vs MIT

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#135]
This issue exists to capture any concerns and discussions about RaptorJIT licensing. I am kicking it off with the concerns already in my own head.

RaptorJIT started from the MIT licensed LuaJIT codebase and then switched to Apache license for new development (#3). This has benefits and potential costs too.

One benefit is that the Apache license explicitly declares that contributions are also licensed the same way (*Submission of Contributions.*) So when you contribute to an Apache licensed project like RaptorJIT you are automatically licensing your code under the Apache license too, and so the licensing status of the project is known. (The MIT license has no such clause and this exposes users to uncertainty about who _really_ owns the copyright and did they _really_ license it as MIT.)

Another benefit is that the Apache license explicitly grants the right to use any relevant patents (*Grant of Patent License.*) This means that the project can accept contributions from large companies like Cisco, Juniper, IBM, Google, etc, without exposing each party to patent litigation by the others.

I see these two factors as positive and also practically important. LuaJIT has contributions from Cisco, Snabb has contributions from Juniper, and the Apache license takes pains to facilitate relaxed and peaceful cooperation while the MIT license does not especially.

There is are potential costs though. The [open source license compatibility flowchart](http://janelia-flyem.github.io/licenses.html) automatically allows MIT code into Apache projects (RaptorJIT can cherry-pick LuaJIT) but it does not automatically allow Apache code into MIT projects (LuaJIT cannot cherry-pick RaptorJIT.) This can potentially be bad, but I don't see it as acute. First, RaptorJIT is focused on new features and LuaJIT will not want to merge these anyway. Second, straight bug fixes are likely to be separately contributed to LuaJIT directly (as I did in LuaJIT/LuaJIT#305). Third, it is plausible that new forks of LuaJIT will be created in the future (e.g. by the gaming community) and we don't know yet what license will work best for them and which they will choose.

TLDR; Apache seems best for RaptorJIT (distributed maintainership), MIT seems best for PUC Lua (centralized development), case could be made either way for LuaJIT.

That is my braindump. Feedback/concerns welcome.

**Comment by szbnwer** [comment]
ive thought about the gpl that it could make incompatibilities around, but ive found that both mit and apache2 can be turned into gpl3 if that comes into the picture. however is there any combination with other frequently used licences that you cant do with apache but can do with mit?

**Comment by lukego** [comment]
@szbnwer Here is a random license compatibility diagram off the internet:

<img width="733" alt="open_licenses" src="https://user-images.githubusercontent.com/13791/33515060-f57505da-d75d-11e7-81db-26772fc204d4.png">

The most important transition in my mind is that Apache cannot be merged into MIT. This could become a problem if there are more LuaJIT forks in the future, and some of them are MIT licensed and would like to cherry-pick changes from RaptorJIT. However, that is still a hypothetical scenario, and it remains to be seen which forks will appear and which licenses they will choose.
